File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/data/SystemDataProvider.java
package org.opencds.cqf.cql.engine.data;

import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;
import org.opencds.cqf.cql.engine.model.BaseModelResolver;
import org.opencds.cqf.cql.engine.runtime.Code;
import org.opencds.cqf.cql.engine.runtime.CqlType;
import org.opencds.cqf.cql.engine.runtime.Date;
import org.opencds.cqf.cql.engine.runtime.DateTime;
import org.opencds.cqf.cql.engine.runtime.Interval;
import org.opencds.cqf.cql.engine.runtime.Quantity;
import org.opencds.cqf.cql.engine.runtime.Time;
import org.opencds.cqf.cql.engine.runtime.Tuple;

public class SystemDataProvider extends BaseModelResolver implements DataProvider {

    @Override
    public Iterable<Object> retrieve(
            String context,
            String contextPath,
            Object contextValue,
            String dataType,
            String templateId,
            String codePath,
            Iterable<Code> codes,
            String valueSet,
            String datePath,
            String dateLowPath,
            String dateHighPath,
            Interval dateRange) {
        throw new IllegalArgumentException("SystemDataProvider does not support retrieval.");
    }

    @Override
    public String getPackageName() {
        return "org.opencds.cqf.cql.engine.runtime";
    }

    @Override
    public void setPackageName(String packageName) {}

    private Field getProperty(Class<?> clazz, String path) {
        try {
            Field field = clazz.getDeclaredField(path);
            return field;
        } catch (NoSuchFieldException e) {
            if (clazz.getSuperclass() != null) {
                Field field = getProperty(clazz.getSuperclass(), path);
                return field;
            }
            throw new IllegalArgumentException(
                    String.format("Could not determine field for path %s of type %s", path, clazz.getSimpleName()));
        }
    }

    static class AccessorKey {
        private final String path;
        private final Class<?> type;

        AccessorKey(String name, Class<?> type) {
            this.path = name;
            this.type = type;
        }

        @Override
        public boolean equals(Object obj) {
            if (obj instanceof AccessorKey) {
                AccessorKey other = (AccessorKey) obj;
                return path.equals(other.path) && type.equals(other.type);
            }
            return false;
        }

        @Override
        public int hashCode() {
            return path.hashCode() + 31 * type.hashCode();
        }
    }

    private static final Map<AccessorKey, Method> readAccessorCache = new HashMap<>();

    private Method getReadAccessor(Class<?> clazz, String path) {
        // Field field = getProperty(clazz, path);
        var accessorKey = new AccessorKey(path, clazz);
        if (readAccessorCache.containsKey(accessorKey)) {
            return readAccessorCache.get(accessorKey);
        }

        String accessorMethodName =
                String.format("%s%s%s", "get", path.substring(0, 1).toUpperCase(), path.substring(1));
        Method accessor = null;
        try {
            accessor = clazz.getMethod(accessorMethodName);
            readAccessorCache.put(accessorKey, accessor);
        } catch (NoSuchMethodException e) {
            return null;
        }
        return accessor;
    }

    private static final Map<AccessorKey, Method> writeAccessorCache = new HashMap<>();

    private Method getWriteAccessor(Class<?> clazz, String path) {
        var accessorKey = new AccessorKey(path, clazz);
        if (writeAccessorCache.containsKey(accessorKey)) {
            return writeAccessorCache.get(accessorKey);
        }

        Field field = getProperty(clazz, path);
        String accessorMethodName =
                String.format("%s%s%s", "set", path.substring(0, 1).toUpperCase(), path.substring(1));
        Method accessor = null;
        try {
            accessor = clazz.getMethod(accessorMethodName, field.getType());
            writeAccessorCache.put(accessorKey, accessor);
            return accessor;
        } catch (NoSuchMethodException e) {
            // If there is no setMethod with the exact signature, look for a signature that would accept a value of the
            // type of the backing field
            for (Method method : clazz.getMethods()) {
                if (method.getName().equals(accessorMethodName) && method.getParameterCount() == 1) {
                    Class<?>[] parameterTypes = method.getParameterTypes();
                    if (parameterTypes[0].isAssignableFrom(field.getType())) {
                        return method;
                    }
                }
            }
            throw new IllegalArgumentException(String.format(
                    "Could not determine accessor function for property %s of type %s", path, clazz.getSimpleName()));
        }
    }

    @Override
    public Object resolvePath(Object target, String path) {
        if (target == null) {
            return null;
        }

        if (target instanceof Tuple) {
            return ((Tuple) target).getElement(path);
        }

        Class<? extends Object> clazz = target.getClass();
        Method accessor = getReadAccessor(clazz, path);
        if (accessor == null) {
            return null;
        }

        try {
            return accessor.invoke(target);
        } catch (InvocationTargetException e) {
            throw new IllegalArgumentException(String.format(
                    "Errors occurred attempting to invoke the accessor function for property %s of type %s",
                    path, clazz.getSimpleName()));
        } catch (IllegalAccessException e) {
            throw new IllegalArgumentException(String.format(
                    "Could not invoke the accessor function for property %s of type %s", path, clazz.getSimpleName()));
        }
    }

    public void setValue(Object target, String path, Object value) {
        if (target == null) {
            return;
        }

        Class<? extends Object> clazz = target.getClass();
        Method accessor = getWriteAccessor(clazz, path);
        try {
            accessor.invoke(target, value);
        } catch (InvocationTargetException e) {
            throw new IllegalArgumentException(String.format(
                    "Errors occurred attempting to invoke the accessor function for property %s of type %s",
                    path, clazz.getSimpleName()));
        } catch (IllegalAccessException e) {
            throw new IllegalArgumentException(String.format(
                    "Could not invoke the accessor function for property %s of type %s", path, clazz.getSimpleName()));
        }
    }

    @Override
    public Class<?> resolveType(Object value) {
        if (value == null) {
            return Object.class;
        }

        return value.getClass();
    }

    @Override
    public Class<?> resolveType(String typeName) {
        switch (typeName) {
            case "Boolean":
                return Boolean.class;
            case "Integer":
                return Integer.class;
            case "Decimal":
                return BigDecimal.class;
            case "String":
                return String.class;
            case "Quantity":
                return Quantity.class;
            case "Interval":
                return Interval.class;
            case "Long":
                return Long.class;
            case "Tuple":
                return Tuple.class;
            case "DateTime":
                return DateTime.class;
            case "Date":
                return Date.class;
            case "Time":
                return Time.class;
            default:
                try {
                    return Class.forName(String.format("%s.%s", getPackageName(), typeName));
                } catch (ClassNotFoundException e) {
                    throw new IllegalArgumentException(
                            String.format("Could not resolve type %s.%s.", getPackageName(), typeName));
                }
        }
    }

    @Override
    public Object createInstance(String typeName) {
        Class<?> clazz = resolveType(typeName);
        try {
            return clazz.getDeclaredConstructor().newInstance();
        } catch (InstantiationException
                | InvocationTargetException
                | ExceptionInInitializerError
                | IllegalAccessException
                | SecurityException
                | NoSuchMethodException e) {
            throw new IllegalArgumentException(
                    String.format("Could not create an instance of class %s.", clazz.getName()));
        }
    }

    @Override
    public Boolean objectEqual(Object left, Object right) {
        if (left == null) {
            return null;
        }

        if (right == null) {
            return null;
        }

        return left.equals(right);
    }

    @Override
    public Boolean objectEquivalent(Object left, Object right) {
        if (left == null && right == null) {
            return true;
        }

        if (left == null) {
            return false;
        }

        if (left instanceof CqlType) {
            return ((CqlType) left).equivalent(right);
        }

        return left.equals(right);
    }

    @Override
    public String resolveId(Object target) {
        return null;
    }

    @Override
    public Object getContextPath(String contextType, String targetType) {
        return null;
    }
}

File: Src/java/engine/src/main/kotlin/org/opencds/cqf/cql/engine/data/SystemDataProvider.kt
package org.opencds.cqf.cql.engine.data

import java.lang.reflect.Field
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.math.BigDecimal
import org.opencds.cqf.cql.engine.model.BaseModelResolver
import org.opencds.cqf.cql.engine.runtime.*
import org.opencds.cqf.cql.engine.runtime.Date

open class SystemDataProvider : BaseModelResolver(), DataProvider {
    override fun retrieve(
        context: String?,
        contextPath: String?,
        contextValue: Any?,
        dataType: String,
        templateId: String?,
        codePath: String?,
        codes: Iterable<Code>?,
        valueSet: String?,
        datePath: String?,
        dateLowPath: String?,
        dateHighPath: String?,
        dateRange: Interval?,
    ): Iterable<Any?>? {
        throw IllegalArgumentException("SystemDataProvider does not support retrieval.")
    }

    @Deprecated("Use packageNames instead")
    override var packageName: String?
        get() = "org.opencds.cqf.cql.engine.runtime"
        set(value) {}

    private fun getProperty(clazz: Class<*>, path: String?): Field {
        try {
            val field = clazz.getDeclaredField(path)
            return field
        } catch (e: NoSuchFieldException) {
            if (clazz.getSuperclass() != null) {
                val field = getProperty(clazz.getSuperclass(), path)
                return field
            }
            throw IllegalArgumentException(
                "Could not determine field for path ${path} of type ${clazz.getSimpleName()}"
            )
        }
    }

    internal class AccessorKey(private val path: String?, private val type: Class<*>) {
        override fun equals(other: Any?): Boolean {
            if (other is AccessorKey) {
                return path == other.path && type == other.type
            }
            return false
        }

        override fun hashCode(): Int {
            return path.hashCode() + 31 * type.hashCode()
        }
    }

    private fun getReadAccessor(clazz: Class<*>, path: String?): Method? {
        // Field field = getProperty(clazz, path);
        val accessorKey = AccessorKey(path, clazz)
        if (readAccessorCache.containsKey(accessorKey)) {
            return readAccessorCache.get(accessorKey)
        }

        val accessorMethodName = "get${path!![0].uppercase()}${path.substring(1)}"
        var accessor: Method? = null
        try {
            accessor = clazz.getMethod(accessorMethodName)
            readAccessorCache.put(accessorKey, accessor)
        } catch (e: NoSuchMethodException) {
            return null
        }
        return accessor
    }

    private fun getWriteAccessor(clazz: Class<*>, path: String?): Method {
        val accessorKey = AccessorKey(path, clazz)
        if (writeAccessorCache.containsKey(accessorKey)) {
            return writeAccessorCache.get(accessorKey)!!
        }

        val field = getProperty(clazz, path)
        val accessorMethodName = "set${path!![0].uppercase()}${path.substring(1)}"
        var accessor: Method? = null
        try {
            accessor = clazz.getMethod(accessorMethodName, field.getType())
            writeAccessorCache.put(accessorKey, accessor)
            return accessor
        } catch (e: NoSuchMethodException) {
            // If there is no setMethod with the exact signature, look for a signature that would
            // accept a value of the
            // type of the backing field
            for (method in clazz.getMethods()) {
                if (method.getName() == accessorMethodName && method.getParameterCount() == 1) {
                    val parameterTypes = method.getParameterTypes()
                    if (parameterTypes[0].isAssignableFrom(field.getType())) {
                        return method
                    }
                }
            }
            throw IllegalArgumentException(
                "Could not determine accessor function for property ${path} of type ${clazz.getSimpleName()}"
            )
        }
    }

    override fun resolvePath(target: Any?, path: String?): Any? {
        if (target == null) {
            return null
        }

        if (target is Tuple) {
            return target.getElement(path)
        }

        val clazz: Class<out Any?> = target.javaClass
        val accessor = getReadAccessor(clazz, path)
        if (accessor == null) {
            return null
        }

        try {
            return accessor.invoke(target)
        } catch (e: InvocationTargetException) {
            throw IllegalArgumentException(
                "Errors occurred attempting to invoke the accessor function for property ${path} of type ${clazz.getSimpleName()}"
            )
        } catch (e: IllegalAccessException) {
            throw IllegalArgumentException(
                "Could not invoke the accessor function for property ${path} of type ${clazz.getSimpleName()}"
            )
        }
    }

    override fun setValue(target: Any?, path: String?, value: Any?) {
        if (target == null) {
            return
        }

        val clazz: Class<out Any?> = target.javaClass
        val accessor = getWriteAccessor(clazz, path)
        try {
            accessor.invoke(target, value)
        } catch (e: InvocationTargetException) {
            throw IllegalArgumentException(
                "Errors occurred attempting to invoke the accessor function for property ${path} of type ${clazz.getSimpleName()}"
            )
        } catch (e: IllegalAccessException) {
            throw IllegalArgumentException(
                "Could not invoke the accessor function for property ${path} of type ${clazz.getSimpleName()}"
            )
        }
    }

    override fun resolveType(value: Any?): Class<*> {
        if (value == null) {
            return Any::class.java
        }

        return value.javaClass
    }

    override fun resolveType(typeName: String?): Class<*> {
        // Important: Boolean, Integer, and Long must be mapped to their Java Object types here
        when (typeName) {
            "Boolean" -> return Boolean::class.javaObjectType
            "Integer" -> return Int::class.javaObjectType
            "Decimal" -> return BigDecimal::class.java
            "String" -> return String::class.java
            "Quantity" -> return Quantity::class.java
            "Interval" -> return Interval::class.java
            "Long" -> return Long::class.javaObjectType
            "Tuple" -> return Tuple::class.java
            "DateTime" -> return DateTime::class.java
            "Date" -> return Date::class.java
            "Time" -> return Time::class.java
            else ->
                try {
                    return Class.forName("${packageName}.${typeName}")
                } catch (e: ClassNotFoundException) {
                    throw IllegalArgumentException(
                        "Could not resolve type ${packageName}.${typeName}."
                    )
                }
        }
    }

    override fun createInstance(typeName: String?): Any? {
        val clazz = resolveType(typeName)
        try {
            return clazz.getDeclaredConstructor().newInstance()
        } catch (e: InstantiationException) {
            throw IllegalArgumentException(
                "Could not create an instance of class ${clazz.getName()}."
            )
        } catch (e: InvocationTargetException) {
            throw IllegalArgumentException(
                "Could not create an instance of class ${clazz.getName()}."
            )
        } catch (e: ExceptionInInitializerError) {
            throw IllegalArgumentException(
                "Could not create an instance of class ${clazz.getName()}."
            )
        } catch (e: IllegalAccessException) {
            throw IllegalArgumentException(
                "Could not create an instance of class ${clazz.getName()}."
            )
        } catch (e: SecurityException) {
            throw IllegalArgumentException(
                "Could not create an instance of class ${clazz.getName()}."
            )
        } catch (e: NoSuchMethodException) {
            throw IllegalArgumentException(
                "Could not create an instance of class ${clazz.getName()}."
            )
        }
    }

    override fun objectEqual(left: Any?, right: Any?): Boolean? {
        if (left == null) {
            return null
        }

        if (right == null) {
            return null
        }

        return left == right
    }

    override fun objectEquivalent(left: Any?, right: Any?): Boolean? {
        if (left == null && right == null) {
            return true
        }

        if (left == null) {
            return false
        }

        if (left is CqlType) {
            return left.equivalent(right)
        }

        return left == right
    }

    override fun resolveId(target: Any?): String? {
        return null
    }

    override fun getContextPath(contextType: String?, targetType: String?): Any? {
        return null
    }

    companion object {
        private val readAccessorCache: MutableMap<AccessorKey?, Method?> =
            HashMap<AccessorKey?, Method?>()

        private val writeAccessorCache: MutableMap<AccessorKey?, Method> =
            HashMap<AccessorKey?, Method>()
    }
}
